File: src/com/mistraltech/smogen/plugin/MatcherGeneratorOptionsDialog.java
package com.mistraltech.smogen.plugin;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.ui.ValidationInfo;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClass;
import com.mistraltech.smogen.codegenerator.matchergenerator.MatcherGeneratorProperties;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

public class MatcherGeneratorOptionsDialog extends DialogWrapper implements MatcherGeneratorOptionsPanelDataSource {
    private static final String RECENTS_KEY = "MatcherGeneratorOptionsDialog.RecentsKey";

    private final Project project;
    private final PsiClass matchedClass;
    private final MatcherGeneratorOptionsPanel matcherGeneratorOptionsPanel;
    private final List<VirtualFile> candidateSourceRoots;
    private final MatcherGeneratorProperties generatorProperties;

    public MatcherGeneratorOptionsDialog(@NotNull Project project,
                                         @NotNull PsiClass matchedClass,
                                         @NotNull List<VirtualFile> candidateSourceRoots,
                                         @NotNull MatcherGeneratorProperties generatorProperties) {
        super(project);

        this.generatorProperties = generatorProperties;
        setModal(true);
        setTitle("Generate Matcher");

        this.project = project;
        this.matchedClass = matchedClass;
        this.candidateSourceRoots = candidateSourceRoots;

        this.matcherGeneratorOptionsPanel = new MatcherGeneratorOptionsPanel(this);

        init();
    }

    @Override
    protected void doOKAction() {
        super.doOKAction();

        final VirtualFile sourceRoot = matcherGeneratorOptionsPanel.getSelectedSourceRoot();

        generatorProperties.setClassName(matcherGeneratorOptionsPanel.getSelectedClassName())
                .setFactoryMethodPrefix(matcherGeneratorOptionsPanel.isAn() ? "an" : "a")
                .setExtensible(matcherGeneratorOptionsPanel.isMakeExtensible())
                .setConcreteSubclassName(matcherGeneratorOptionsPanel.getConcreteSubclassName())
                .setSuperClassName(matcherGeneratorOptionsPanel.getSuperClassName())
                .setPackageName(matcherGeneratorOptionsPanel.getSelectedPackageName())
                .setSourceRoot(sourceRoot);
    }

    @Nullable
    @Override
    protected ValidationInfo doValidate() {
        ValidationInfo validationInfo = matcherGeneratorOptionsPanel.doValidate();

        if (validationInfo != null) {
            return validationInfo;
        }

        return super.doValidate();
    }

    @Nullable
    protected JComponent createCenterPanel() {
        return matcherGeneratorOptionsPanel.getRoot();
    }

    @NotNull
    @Override
    public Project getProject() {
        return project;
    }

    @NotNull
    @Override
    public String getRecentsKey() {
        return RECENTS_KEY;
    }

    @NotNull
    @Override
    public String getPackageName() {
        return generatorProperties.getPackageName();
    }

    @NotNull
    @Override
    public PsiClass getMatchedClass() {
        return matchedClass;
    }

    @NotNull
    @Override
    public String getDefaultClassName() {
        return generatorProperties.getClassName();
    }

    @NotNull
    @Override
    public VirtualFile getDefaultRoot() {
        return generatorProperties.getSourceRoot();
    }

    @NotNull
    @Override
    public List<VirtualFile> getCandidateRoots() {
        return candidateSourceRoots;
    }

    @Override
    public boolean getDefaultIsExtensible() {
        return generatorProperties.isExtensible();
    }
}

File: src/main/kotlin/com/mistraltech/smogen/plugin/MatcherGeneratorOptionsDialog.kt
package com.mistraltech.smogen.plugin

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiClass
import com.mistraltech.smogen.codegenerator.matchergenerator.MatcherGeneratorProperties
import javax.swing.JComponent

class MatcherGeneratorOptionsDialog(
    override val project: Project,
    override val matchedClass: PsiClass,
    override val candidateRoots: List<VirtualFile>,
    private val generatorProperties: MatcherGeneratorProperties
) : DialogWrapper(project), MatcherGeneratorOptionsPanelDataSource {

    override val recentsKey: String = "MatcherGeneratorOptionsDialog.RecentsKey"

    private val matcherGeneratorOptionsPanel: MatcherGeneratorOptionsPanel

    override fun doOKAction() {
        super.doOKAction()

        val sourceRoot = matcherGeneratorOptionsPanel.selectedSourceRoot

        generatorProperties.setClassName(matcherGeneratorOptionsPanel.selectedClassName)
            .setFactoryMethodPrefix(if (matcherGeneratorOptionsPanel.isAn) "an" else "a")
            .setExtensible(matcherGeneratorOptionsPanel.isMakeExtensible)
            .setMatcherSuperClassName(matcherGeneratorOptionsPanel.superClassName)
            .setPackageName(matcherGeneratorOptionsPanel.selectedPackageName)
            .setGenerateInterface(matcherGeneratorOptionsPanel.isGenerateInterface)
            .setSourceRoot(sourceRoot)
    }

    override fun doValidate(): ValidationInfo? {
        val validationInfo = matcherGeneratorOptionsPanel.doValidate()
        return validationInfo ?: super.doValidate()
    }

    override fun createCenterPanel(): JComponent {
        return matcherGeneratorOptionsPanel.root
    }

    override val packageName: String
        get() = generatorProperties.packageName!!
    override val defaultClassName: String
        get() = generatorProperties.className!!
    override val defaultRoot: VirtualFile
        get() = generatorProperties.sourceRoot!!
    override val defaultIsExtensible: Boolean
        get() = generatorProperties.isExtensible

    init {
        isModal = true
        title = "Generate Matcher"
        matcherGeneratorOptionsPanel = MatcherGeneratorOptionsPanel(this)
        init()
    }
}
